base_layer/wallet/src/output_manager_service/service.rs

```diff
@@ -450,46 +450,120 @@ where
     pub async fn query_unspent_outputs_status(
         &mut self,
         utxo_query_timeout_futures: &mut FuturesUnordered<BoxFuture<'static, u64>>,
+        specified_outputs: Option<Vec<Vec<u8>>>,
     ) -> Result<u64, OutputManagerError>
     {
         match self.base_node_public_key.as_ref() {
             None => Err(OutputManagerError::NoBaseNodeKeysProvided),
             Some(pk) => {
-                let unspent_outputs: Vec<DbUnblindedOutput> = self.db.get_unspent_outputs().await?;
-                let mut output_hashes = Vec::new();
-                for uo in unspent_outputs.iter() {
-                    let hash = uo.hash.clone();
-                    output_hashes.push(hash.clone());
-                }
-                let request_key = OsRng.next_u64();
+                let mut first_request_key = 0;
+                let mut unspent_outputs: Vec<Vec<u8>> = if let Some(hashes) = specified_outputs {
+                    hashes
+                } else {
+                    self.db
+                        .get_unspent_outputs()
+                        .await?
+                        .iter()
+                        .map(|uo| uo.hash.clone())
+                        .collect()
+                };
 
-                let request = BaseNodeRequestProto::FetchUtxos(BaseNodeProto::HashOutputs {
-                    outputs: output_hashes.clone(),
-                });
+                // Determine how many rounds of base node request we need to query all the outputs in batches of
+                // max_utxo_query_size
+                let rounds =
+                    ((unspent_outputs.len() as f32) / (self.config.max_utxo_query_size as f32 + 0.1)) as usize + 1;
```

Comment:
Maybe use `ceil()` here? Otherwise, wouldn't you get an extra round if number if `unspent_outputs.len()` is a multiple of `self.config.max_utxo_query_size`? 
```suggestion
                    ((unspent_outputs.len() as f32) / (self.config.max_utxo_query_size as f32 + 0.1)).ceil() as usize;
```

Reply:
Could actually use ceil but it will be the same, there is a sneaky +0.1 in the denominator to avoid the issue you described.